### PR TITLE
feat: Add autodiscoverExclusions global config option

### LINKED ISSUE #21347: Add autodiscoverExclusions to exclude projects fetched by autodiscoverFilter
### What would you like Renovate to be able to do?

I would like to be able to add the whole GitLab group to autodiscoverFilter, yet exclude some legacy projects in that group.

The following approch doesn't work:
```
  autodiscover: true,
  autodiscoverFilter: [
    "whole-group1/*",
    "!/whole-group1/legacy-project-to-skip/",
    ...
```

### If you have any ideas on how this should be implemented, please tell us here.

As suggested in https://github.com/renovatebot/renovate/discussions/21341, it might be possible to achieve with a separate `autodiscoverExclusions`, e.g.

```
  autodiscover: true,
  autodiscoverFilter: [
    "whole-group1/*",
  ],
  autodiscoverExclusions: [
    "whole-group1/legacy-project-to-skip"
  ]
```

The elements in `autodiscoverExclusions` should be subtracted from the projects resolved by `autodiscoverFilter`. Probably it would be good to have an ability to use also regexps in `autodiscoverExclusions` (e.g. `"whole-group1/legacy-project*"` or `"/whole-group1/legacy-group/**/"` or `"/**/.*-static-pages/"`).

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by markhoward7**
## Changes

Add `autodiscoverExclusions` as a global config option that allows users to exclude repositories that would otherwise match the `autodiscoverFilter`.

## Context

Closes #21347 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/global/autodiscover.ts [hunk 1]
```diff
@@ -51,10 +51,15 @@ export async function autodiscoverRepositories(
...
-    );
+    const addFilters = is.string(autodiscoverFilter)
+      ? [autodiscoverFilter]
+      : autodiscoverFilter;
+    const removeFilters = autodiscoverExclusions
+      ? is.string(autodiscoverExclusions)
+        ? [autodiscoverExclusions]
+        : autodiscoverExclusions
+      : [];
+    discovered = applyFilters(discovered, addFilters, removeFilters);
```
Comment (secustor) [inline lib/workers/global/autodiscover.ts:62]:
```suggestion
    const addFilters = prepareFilter(autodiscoverFilter)
    const removeFilters = prepareFilter(autodiscoverExclusions)
    discovered = applyFilters(discovered, addFilters, removeFilters);
```

extract this into a separate function for readability. 

```typescript
function prepareFilter(filter: string | string[] | undefined): string[] {
  if ((is.array(filter))) {
    return filter
  }
  if (is.string(filter)) {
    return [filter]
  }
  return []
}
```

File: lib/workers/global/autodiscover.ts [hunk 2]
```diff
@@ -95,10 +100,13 @@ export async function autodiscoverRepositories(
...
-export function applyFilters(repos: string[], filters: string[]): string[] {
-  const matched = new Set<string>();
-
-  for (const filter of filters) {
+export function applyFilters(
+  repos: string[],
+  addFilters: string[],
+  removeFilters: string[],
+): string[] {
+  const applyFilter = (repos: string[], filter: string): string[] => {
```
Comment (secustor) [inline lib/workers/global/autodiscover.ts:108]: Do not use local functions 

https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#functions

File: lib/workers/global/autodiscover.ts [hunk 3]
```diff
@@ -112,6 +120,13 @@ export function applyFilters(repos: string[], filters: string[]): string[] {
...
-  return [...matched];
+    return [...matched];
+  };
+
+  const matched = addFilters.flatMap((filter) => applyFilter(repos, filter));
+  const removeMatched = removeFilters.flatMap((filter) =>
+    applyFilter(matched, filter),
+  );
+
+  return [...matched.filter((matched) => !removeMatched.includes(matched))];
```
Comment (secustor) [inline lib/workers/global/autodiscover.ts:131]: This results in potential duplicates

**Review by secustor — CHANGES_REQUESTED** [review]
Thanks for your contribution!

**Comment by markhoward7** [comment]
Accidentally used another account, but created the corresponding discussion #25812

**Pull request closed without merging**